cosmwasm/osmosis-tokenfactory-token-minter/src/contract.rs

```diff
@@ -0,0 +1,315 @@
+
+                    vec![
+                        CosmosMsg::Custom(TokenFactoryMsg::CreateDenom {
+                            subdenom: subdenom.to_owned(),
+                        }),
+                        #[allow(deprecated)]
+                        CosmosMsg::Stargate {
+                            type_url: MsgSetDenomMetadata::type_url(),
+                            value: MsgSetDenomMetadata {
+                                sender: env.contract.address.to_string(),
+                                metadata: Some(protos::cosmos::bank::v1beta1::Metadata {
+                                    description: "".to_string(),
+                                    denom_units: vec![
+                                        DenomUnit {
+                                            denom: denom.clone(),
+                                            exponent: 0,
+                                            aliases: vec![metadata.symbol.clone()],
+                                        },
+                                        DenomUnit {
+                                            denom: metadata.symbol.clone(),
+                                            exponent: metadata.decimals.into(),
+                                            aliases: vec![],
+                                        },
+                                    ],
+                                    base: denom.clone(),
+                                    display: metadata.symbol.clone(),
+                                    name: metadata.name,
+                                    symbol: metadata.symbol,
+                                    uri: "".to_string(),
+                                    uri_hash: "".to_string(),
+                                }),
+                            }
+                            .encode_to_vec()
+                            .into(),
+                        },
+                        CosmosMsg::Custom(TokenFactoryMsg::ChangeAdmin {
+                            denom,
+                            new_admin_address: TOKEN_ADMIN.load(deps.storage)?.to_string(),
+                        }),
+                    ]
+                }
+                WrappedTokenMsg::MintTokens {
+                    denom,
+                    amount,
+                    mint_to_address,
+                } => {
+                    vec![CosmosMsg::Custom(TokenFactoryMsg::MintTokens {
+                        denom,
+                        amount,
+                        mint_to_address,
+                    })]
+                }
+                WrappedTokenMsg::BurnTokens {
+                    denom,
+                    amount,
+                    burn_from_address,
+                    ..
+                } => {
+                    let contains_base_token = info
+                        .funds
+                        .iter()
+                        .any(|coin| coin.denom == denom && coin.amount == amount);
```

Comment:
are funds automatically sent to the contract? or do we need to take them somehow?